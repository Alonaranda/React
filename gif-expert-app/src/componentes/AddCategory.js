import React, {useState} from 'react';
import PropTypes from 'prop-types';

export const AddCategory = ({setCategories}) => {

    const [inputValue, setValue] = useState('');

    const handleInputText = (e) => {
        setValue(e.target.value);
        //console.log('Change', inputValue);
    }

    const handleSubmit = (e) => {
        e.preventDefault();
        //console.log('Submit!', inputValue);
        if(inputValue.trim().length > 2){
            setCategories(cate => [inputValue, ...cate, ]);
            setValue('');
        }
    }

    return (
        <form onSubmit={handleSubmit}>
            <p>{inputValue}</p>
            <input
                type="text"
                value={inputValue}
                onChange={handleInputText}
                placeholder='Buscado tu Gif'
            ></input>
        </form>
        
    )
}

//PropTypes
AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
