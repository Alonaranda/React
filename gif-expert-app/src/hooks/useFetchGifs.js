import {useState,useEffect} from 'react';
import {getGifs} from '../helpers/getGifs';

//Son como functional component

export const useFetchGifs= (category) => {
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    useEffect(()=>{
        getGifs(category).then(imgs =>{setState({
            data:imgs,
            loading: false
        })})
    },[category])

    return state;
}