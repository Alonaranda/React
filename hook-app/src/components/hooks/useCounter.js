import {useState} from 'react'


export const useCounter = (initialState = 10) => {

    //usamos el useState
    const [state, setstate] = useState(initialState);

    //Incrementar
    const increment = () => {
        setstate(state + 1);
    }

    //Decrementar
    const decrement = (factor = 1) =>{
        setstate(state - factor);
    } 

    //Reset
    const reset = () => {
        setstate(initialState);
    }

    return {
        state,
        increment,
        decrement, 
        reset
    };
}
