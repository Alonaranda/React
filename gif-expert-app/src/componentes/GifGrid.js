import React from "react";
import { useFetchGifs } from "../hooks/useFetchGifs";
import { GifGridItem } from "./GifGridItem";
import PropTypes from 'prop-types';
//import { getGifs } from "../helpers/getGifs";


export const GifGrid = ({ category }) => {
  //Al ejecutar una funcion dentro de una api, se vuelve un ciclo infinito
  //UseEffect hace que no hagamos ciclo infinito
  //Ejecuta cierto codigo de manera condicional
  //El arreglo vacio, indica que solo se ejecuta una vez

  
    
  //const [images, setImages] = useState([]);
    //Usamos helpers
    const {data:images, loading} = useFetchGifs(category);
    //console.log(data, loading);

    /*
    useEffect(() => {
        getGifs(category)
            .then(imgs =>{
                setImages(imgs);
        })
    }, [category]);//Si la category cambia, se vuelve a ejecutar
    */
  

  return (
    <>
      <h3 className="animate_animated animate_fadeIn">{category}</h3>
      {loading && <p className="animate_animated animate_flash">Loading...</p>}

      <div className="card-grid">
        {images.map(img => (
          <GifGridItem key={img.id} {...img} />
        ))}
      </div>
    </>
  );
};

GifGrid.propTypes = {
  category: PropTypes.string.isRequired
}
