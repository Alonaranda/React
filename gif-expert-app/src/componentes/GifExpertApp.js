import React, {useState} from 'react'
import { AddCategory } from './AddCategory';
import { GifGrid } from './GifGrid';

export const GifExpertApp = ({defaultCategories=[]}) => {
    //const categories = ['Inuyasha', 'Naruto', 'Samurai X'];
    const [categories, setCategories] = useState(defaultCategories);
    /*
    const handleAdd = () => {
        //setCategories([...categories, 'DeathNote']);
        setCategories( cate => [...cate, 'DeathNote']);
    }*/

    //Para pasar datos en el <AddCategory/> agregamos que vamos a utilizar
    return (
        <>
            <h2>GifExpertApp</h2>
             
            <AddCategory setCategories={setCategories}/>
            <hr/>
            <ol>
                {
                    categories.map(category => 
                        <GifGrid 
                        key={category}
                        category={category}
                        />
                    //<li key={category}>{category}</li>
                    ) //El key no debe ser el indice, casi siempre es el id de la BD
                }
            </ol>
        </>
    )
}

