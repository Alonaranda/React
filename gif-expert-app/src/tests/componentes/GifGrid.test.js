import React from 'react'
import {shallow} from 'enzyme';
import '@testing-library/jest-dom';
import { GifGrid } from '../../componentes/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
import { wrap } from 'module';
jest.mock('../../hooks/useFetchGifs');

describe('Probando GifGrid', () => {
    const category = 'Inuyasha';

    //Hacer un snapshot
    test('Hacer un snapshot ', () => {
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });
        const wrapper = shallow(<GifGrid category={category}/>)
        expect(wrapper).toMatchSnapshot();
    });

    //UseFetch ya me regresa lka informacion 
    //Debe de mostrar items cuando se cargan imagenes
    //Un mock es fingir algo
    test('Debe de mostrar items cuando carga useFetchGifs ', () => {

        const gifs = [{
            id: 'ABC',
            url: 'https://localhost/cosa.jpg',
            title: 'Titulo'
        },
        {
            id: 'ASD',
            url: 'https://localhost/cosa.jpg',
            title: 'Titulo'
        },
    ]

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });
        const wrapper = shallow(<GifGrid category={category}/>)
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('p').exists()).toBe(false);
        expect(wrapper.find('GifGridItem').length).toBe(gifs.length)
    });

});