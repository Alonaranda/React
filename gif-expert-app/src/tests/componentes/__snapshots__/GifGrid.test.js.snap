// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Probando GifGrid Debe de mostrar items cuando carga useFetchGifs  1`] = `
<Fragment>
  <h3
    className="animate_animated animate_fadeIn"
  >
    Inuyasha
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="Titulo"
      url="https://localhost/cosa.jpg"
    />
    <GifGridItem
      id="ASD"
      key="ASD"
      title="Titulo"
      url="https://localhost/cosa.jpg"
    />
  </div>
</Fragment>
`;

exports[`Probando GifGrid Hacer un snapshot  1`] = `
<Fragment>
  <h3
    className="animate_animated animate_fadeIn"
  >
    Inuyasha
  </h3>
  <p
    className="animate_animated animate_flash"
  >
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
